### PR TITLE
inventory control stream

### PR DESCRIPTION
**Pull request by fspmarshall**
This PR sets up some basic features that are prerequisites for upcoming upgrade system work:

1. Added a new "instance" cert type which encodes all system roles of an instance (previously all certs were specific to a service).
2. Added an instance-level bidirectional GRPC stream using the new cert type that allows an instance to heartbeat information about all aspects of itself (as opposed to requiring separate certs/clients per-service), and allows the auth server to send commands to individual instances directly.
3. Ported SSH Server heartbeats to use the new instance stream.

The above changes are all about moving teleport toward having a more unified view of its inventory, and making it easier to have auth servers and `tctl` directly affect changes against non-auth instances.

note: still doing some polish and more tests incoming.

### CODE REVIEW

File: api/client/inventory.go [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+// between teleport instances and auth servers.
+type UpstreamInventoryControlStream interface {
+	// Send attempts to send a downstream message.  An error returned from this
+	// method either indicates that the stream itself has failed, or that the
+	// supplied context was canceled.
+	Send(ctx context.Context, msg proto.DownstreamInventoryMessage) error
+	// Recv access the incoming/upstream message channel.
+	Recv() <-chan proto.UpstreamInventoryMessage
+	// Close closes the underlying stream with optional error.
+	Close(...error)
```
Comment (zmb3) [inline api/client/inventory.go:58]: Why variadic here? We can handle an optional error with a single argument that is either nil or not.

I would also call this `CloseWithError` (there's precedent, see `io.PipeWriter` for example)

File: api/client/inventory.go [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+	// Recv accesses the incoming/downstream message channel.
+	Recv() <-chan proto.DownstreamInventoryMessage
+	// Close closes the underlying stream.
+	Close(...error)
+	// Done signals that the stream has been closed.
+	Done() <-chan struct{}
+	// Error checks for any error associated with stream closure (returns `nil` if
+	// the stream is open, or if the the stream was closed explicitly via the Close()
+	// method).
+	Error() error
```
Comment (zmb3) [inline api/client/inventory.go:45]: So this interface offers no way to tell whether the stream is still open or whether it was closed successfully? If so, that seems like a confusing API.

Could the `Close` method return an error and we can use that to indicate whether the close operation succeeded?

File: api/client/inventory.go [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+	// the stream is open, or if the stream was closed explicitly via the Close()
+	// method).
+	Error() error
+}
+
+// InventoryControlStreamPipe creates the two halves of an inventory control stream over an in-memory
+// pipe.
+func InventoryControlStreamPipe() (UpstreamInventoryControlStream, DownstreamInventoryControlStream) {
+	pipe := &pipeControlStream{
+		downC: make(chan proto.DownstreamInventoryMessage, 1),
```
Comment (zmb3) [inline api/client/inventory.go:71]: I'm always suspicious of a buffered channel of size 1 - can you elaborate on why this is the case?

File: api/client/inventory.go [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+		case i.recvC <- msg:
+		case <-i.Done():
+			// stream closed by other goroutine
+			return
+		}
+	}
+}
+
+// runSendLoop pulls messages off of the sendC channel, applies the appropriate protobuf wrapper types,
+// and sends them over the channel.
```
Comment (zmb3) [inline api/client/inventory.go:240]:
```suggestion
// and sends them over the stream.
```

File: api/types/system_role.go [hunk 5]
```diff
@@ -60,6 +60,9 @@ const (
 	RoleWindowsDesktop SystemRole = "WindowsDesktop"
 	// RoleBot is a role for a bot.
 	RoleBot SystemRole = "Bot"
+	// RoleInstance is a role implicitly held by teleport instances, and is used as a
```
Comment (zmb3) [inline api/types/system_role.go:63]: What is a Teleport "instance"? A process?

**Review by zmb3 — COMMENTED** [review]
I realize this is a draft, but had a few minutes and took a look. It's shaping up to be a pretty big change. I ran out of time and didn't get through the whole thing, but added a few comments.

In general, for a feature that's so critical to get right, I would prefer to see smaller changes that are easier to review, and better test coverage. As someone not intimately familiar with the details, a lot of this is hard for me to reason about.

File: api/client/inventory.go (hunk 3 above)
Comment (fspmarshall) [reply api/client/inventory.go]: Was using a buffer size of 1 here to keep the amount of slop as low as possible while still supporting simultaneous sending of "hello" messages (both sides of the stream were synchronously sending their own hello message out before trying to recv the incoming hello). Changed to having auth wait for client hello before sending its reply. Pros and cons to doing it either way, but above channels are now unbuffered.

File: api/types/system_role.go (hunk 5 above)
Comment (fspmarshall) [reply api/types/system_role.go:63]: That's the million dollar question lol.  Given how graceful restarts work, its technically allowable for multiple teleport processes with the same ID to be running simultaneously.  In this case I'm using the term "instance" to try to refer to the abstract identity represented by an id/data_dir.  Names are hard.

Comment (zmb3) [reply api/types/system_role.go:63]: That explanation makes sense. I would add it right to the godoc.

File: lib/inventory/store.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+func (s *shard) remove(handle UpstreamHandle) {
+	s.rw.Lock()
+	defer s.rw.Unlock()
+	e, ok := s.m[handle.Hello().ServerID]
+	if !ok {
+		return
+	}
+	for i, h := range e.handles {
+		if handle == h {
+			e.handles = append(e.handles[:i], e.handles[i+1:]...)
```
Comment (espadolini) [inline lib/inventory/store.go:168]: `swap_remove`?

File: lib/service/connect.go [hunk 7]
```diff
@@ -49,6 +50,8 @@ import (
 // reconnectToAuthService continuously attempts to reconnect to the auth
 // service until succeeds or process gets shut down
 func (process *TeleportProcess) reconnectToAuthService(role types.SystemRole) (*Connector, error) {
+	// TODO(fspmarshall): we should probably have a longer retry period for Instance certs
+	// in order to avoid catastrophic load in the event of an auth server downgrade.
```
Comment (espadolini) [inline lib/service/connect.go:54]: Auth downgrade without downgrading clients first is not a supported configuration tho 🤔

File: lib/srv/heartbeatv2.go [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+	if c.pollInterval == 0 {
+		c.pollInterval = defaults.HeartbeatCheckPeriod
+	}
+	if c.fallbackBackoff == 0 {
+		// only set externally during tests
+		c.fallbackBackoff = time.Minute
+	}
+}
+
+func (h *HeartbeatV2) run() {
```
Comment (espadolini) [inline lib/srv/heartbeatv2.go:166]: Refactor into more easily digested functions?

**Review by espadolini — COMMENTED** [review]
First pass, wasn't very systematic because I tried to follow the execution flow.

File: lib/srv/heartbeatv2_test.go [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+		handle:      handle,
+		streamC:     streamC,
+		pollChanged: 1, // first poll is always true
+	}
+}
+
+// TestHeartbeatV2Basics verifies the basic functionality of HeartbeatV2 under various conditions by
+// injecting a fake driver and attempting to force the HeartbeatV2 into all of its happy and unhappy
+// states while monitoring test events to verify expected behaviors.
+func TestHeartbeatV2Basics(t *testing.T) {
```
Comment (zmb3) [inline lib/srv/heartbeatv2_test.go:145]: Can this test be run in parallel? It takes 5-6 seconds on my machine which is a bit heavy for a unit test.

File: lib/srv/heartbeatv2_test.go [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// set up fake hb driver that lets us easily inject failures for
+	// the diff steps and assists w/ faking inventory control handles.
+	driver := newFakeHeartbeatDriver()
+	defer driver.handle.Close() // driver includes fake handle that needs cleanup
```
Comment (zmb3) [inline lib/srv/heartbeatv2_test.go:152]: For test utilities, I prefer they clean up after themselves. We've had a lot of flaky tests due to missing or incorrect cleanup.

Something like:

```
driver := newFakeHeartbeatDriver(t)
// newFakeHeartBeatDriver calls t.Cleanup for any cleanup actions

// no need to defer a close, cleanup happens automatically when test is finished
```

File: lib/srv/heartbeatv2_test.go [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+
+	hb := newHeartbeatV2(driver.handle, driver, heartbeatV2Config{
+		announceInterval: time.Millisecond * 200,
+		pollInterval:     time.Millisecond * 50,
+		fallbackBackoff:  time.Millisecond * 400,
+		testEvents:       make(chan hbv2TestEvent, 1028),
+	})
+	go hb.Run()
```
Comment (zmb3) [inline lib/srv/heartbeatv2_test.go:160]: Any reason we can't have `hb.Run()` take a context? It would be nice to standardize and use context for the cancelation of any long-running operation.

File: lib/srv/heartbeatv2_test.go [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+	awaitEvents(t, hb.testEvents,
+		expect(hbv2FallbackOk, hbv2FallbackOk),
+		deny(hbv2FallbackErr, hbv2AnnounceOk, hbv2AnnounceErr),
+	)
+
+	// make a stream available to the heartbeat instance
+	// (note: we don't need to pull from our half of the stream since
+	// fakeHeartbeatDriverInner doesn't actually send any messages across it).
+	stream := driver.newStream(ctx, t)
+	defer stream.Close()
```
Comment (zmb3) [inline lib/srv/heartbeatv2_test.go:203]: Same thing here, can `newStream` use `t.Cleanup` so we don't need to remember to close it?

File: lib/auth/auth.go [hunk 13]
```diff
@@ -554,6 +568,7 @@ func (a *Server) updateVersionMetrics() {
 
 func (a *Server) Close() error {
 	a.cancelFunc()
+	a.inventory.Close()
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/auth/auth.go:571]: I don't feel too strongly either way about the `inventory.Controller` in particular, but I've come around to the perspective that explicit close is better for persistent (i.e. not request scoped) resources/components.  Not sure I'm *right* about that, but here are some of my thoughts on it (in no particular order):

- Its nice to be able to control the order in things close and/or be able to easily inject things that you might want to happen "before" a specific component starts to close.  In particular, it is much easier to introduce a "graceful shutdown" mode later down the line if closure is compartmentalized from the start, rather than closure basically happening at every level all at once.
- Its generally easier to switch from explicit close to implicit close than it is to switch from implicit close to explicit close, so if you're not sure which one you should use, seems best to default to explicit.
- In the broader go ecosystem contexts are almost never used for the closure of long-lived resources.  Not necessarily a good reason not to do it, but I can think of at least two reason for why explicit closure seems to be preferred by most folks over implicit closure:
  1. Implicit closure means that you might never actually get to do what you want to do as part of your particular closure process.  Once main exits, all background goroutines terminate without even being able to run deferred functions.  Any component that relies on context cancellation to begin its cleanup process may simply never have its cleanup run.  This is fine for a lot of components, but implicit closure is contagious.  If any component "above" you is implicitly closed, then you are in effect also implicitly closed.  This isn't a disaster... programs exit unexpectedly sometimes and should be able to handle that, but it still seems that we aught to try to respect Close() methods where we can.
  2. Implicit/asynchronous closure is incompatible with finite resources. E.g. if a component opens a listener on some port P, its reasonable to assume that two of said component can't be run at the same time, but one aught to feel safe in the assumption that once Close() has been called on the first instance, starting a second should now succeed. If all cleanup is asynchronous (and, as mentioned above, implicit/asynchronous closure is contagious) then that can't be relied on.
  
Like I said... I don't think it really matters in this case, but for the above reasons I've been trying to default to explicit close unless I can think of some specific reason why implicit close is better in the particular case I'm looking at.

File: lib/auth/auth.go [hunk 14]
```diff
@@ -2137,11 +2152,21 @@ func (a *Server) GenerateHostCerts(ctx context.Context, req *proto.HostCertsRequ
...
+	}
+
+	if req.Role == types.RoleInstance && len(systemRoles) == 0 {
+		return nil, trace.BadParameter("cannot generate instance cert with no system roles")
+	}
+
 	// generate host TLS certificate
 	identity := tlsca.Identity{
 		Username:        HostFQDN(req.HostID, clusterName.GetClusterName()),
 		Groups:          []string{req.Role.String()},
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/auth/auth.go:2167]: Yes, but more specifically, I'd like code that doesn't know what to do with a multirole cert to not fail in some odd or subtle/"almost right" way.  If a multirole cert can be misinterpreted as a single-role cert, that could result in confusing and difficult to diagnose behavior.  Much better IMO to avoid the potential ambiguity.

File: lib/service/connect.go (hunk 7 above)
Comment (fspmarshall) [reply lib/service/connect.go:54]: Having this particular feature break with the norm and be back-combat with older servers was a deliberate deviation RJ and I decided to make.  Basically, we thought the potential negative fallout of heartbeats being totally non-functional if someone needed to revert their auth after a V10 upgrade was a bit too extreme.  While we don't officially support older auths, teleport usually "kinda works" in that scenario.  Cache might go into an error loop, and some specific features might be pretty buggy, but usually you could at least theoretically still ssh into nodes (even if w/ very poor perf).

This isn't unprecedented (e.g. a lot of the GRPC methods that used to be normal HTTP methods kept the old HTTP method as a fallback for at least one version).

File: lib/auth/auth.go [hunk 15]
```diff
@@ -2175,6 +2200,49 @@ func (a *Server) GenerateHostCerts(ctx context.Context, req *proto.HostCertsRequ
...
+func (a *Server) UnstableGetSystemRoleAssertions(ctx context.Context, serverID string, assertionID string) (proto.UnstableSystemRoleAssertionSet, error) {
+	set, err := a.unstable.GetSystemRoleAssertions(ctx, serverID, assertionID)
+	return set, trace.Wrap(err)
+}
+
+func (a *Server) RegisterInventoryControlStream(ics client.UpstreamInventoryControlStream, hello proto.UpstreamInventoryHello) {
+	a.inventory.RegisterControlStream(ics, hello)
+}
+
+func (a *Server) GetInventoryStatus(ctx context.Context, req proto.InventoryStatusRequest) proto.InventoryStatusSummary {
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/auth/auth.go:2216]: This method was added to support the `tctl inventory status` command.  Added some comments there to better explain the planned future, but basically I'd like for the *default* behavior of getting the inventory status to show cluster-wide status information (e.g. you have 3 nodes on v1.2.3, 5 nodes on v2.3.4, etc...).  The `--connected` flag shows you information about the actual connections that the auth server you are directly connected to has.  Good for pings and debug, but not a very useful default.

Basically, `Connected == false` will be the default behavior... it just so happens that the default behavior doesn't do anything yet.

File: lib/auth/auth_with_roles.go [hunk 16]
```diff
@@ -536,18 +536,167 @@ func (a *ServerWithRoles) GenerateHostCerts(ctx context.Context, req *proto.Host
...
+		return trace.AccessDenied("cannot assert non-service system role %q", req.SystemRole)
+	}
+
+	return a.authServer.UnstableAssertSystemRole(ctx, req)
+}
+
+func (a *ServerWithRoles) RegisterInventoryControlStream(ics client.UpstreamInventoryControlStream) error {
+	// Ensure that caller is a teleport server
+	role, ok := a.context.Identity.(BuiltinRole)
+	if !ok || !role.IsServer() {
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/auth/auth_with_roles.go:629]: Refactored but ended up reverting.  Turns out that we almost always need to perform this typecast anyway since we proceed to access fields that are unique to `BuiltinRole`, so a helper function didn't actually reduce verbosity at all, and increased the amount of typecasts that we had to perform.

File: lib/auth/auth_with_roles.go [hunk 17]
```diff
@@ -536,18 +536,167 @@ func (a *ServerWithRoles) GenerateHostCerts(ctx context.Context, req *proto.Host
...
+	upstreamHello.Services = filteredServices
+
+	// send downstream hello (note: in theory we could send hellos simultaneously to slightly
+	// improve perf, but there is a potential benefit to having the downstream hello serve
+	// double-duty as an indicator of successful auth).
+	downstreamHello := proto.DownstreamInventoryHello{
+		Version:  teleport.Version,
+		ServerID: a.authServer.ServerID,
+	}
+	if err := ics.Send(a.CloseContext(), downstreamHello); err != nil {
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/auth/auth_with_roles.go:674]: Passing the rpc-scoped context back into the stream would be redundant I think (the context associated with the GRPC stream signals the stream's closure, so we'd sort of just be waiting on the same cancellation event from two separate angles).  IMO passing the auth server's CloseContext here is also a bit redundant, since it should close the GRPC server if it is closing, but I think its the less redundant of the two options.

File: api/client/inventory.go [hunk 18]
```diff
@@ -0,0 +1,510 @@
...
+// type, and pushes them to the recvC channel.
+func (i *upstreamICS) runRecvLoop(stream proto.AuthService_InventoryControlStreamServer) {
+	for {
+		oneOf, err := stream.Recv()
+		if err != nil {
+			// preserve eof errors
+			if !trace.IsEOF(err) {
+				err = trace.Errorf("inventory control stream recv failed: %v", trail.FromGRPC(err))
+			}
+			i.CloseWithError(err)
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply api/client/inventory.go:393]: @zmb3 requested that we go with standard closer method signatures here.  While there is much less of an established patter for `CloseWithError` than for `Close`, it does seem to typically return an error value (e.g. [`io.PipeReader.CloseWithError`](https://pkg.go.dev/io#PipeReader.CloseWithError)).

File: lib/inventory/controller.go [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+	store           *Store
+	auth            Auth
+	serverKeepAlive time.Duration
+	serverTTL       time.Duration
+	testEvents      chan testEvent
+	closeContext    context.Context
+	cancel          context.CancelFunc
+}
+
+func NewController(auth Auth, opts ...controllerOption) *Controller {
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/inventory/controller.go]: The options are private on purpose, since they are only used in local tests.  Renamed them to `controllerTestOption` to better clarify this.  In terms of why go with this instead of a struct, adding an additional unhappy path and losing compile-time input validation for a single public argument seemed unnecessary.

File: lib/inventory/controller.go [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+// one handle is selected pseudorandomly to be observed.
+func (c *Controller) Iter(fn func(UpstreamHandle)) {
+	c.store.Iter(fn)
+}
+
+func (c *Controller) testEvent(event testEvent) {
+	if c.testEvents == nil {
+		return
+	}
+	c.testEvents <- event
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/inventory/controller.go:140]: Since this is only used in tests, I prefer the blocking.  Definitely a personal taste thing, but if the events aren't being consumed for some reason, I'd rather see the test hit an explicit timeout rather than lose events.  Feels like that might result in unpredictable behavior that could be more work to debug.

I could add some mechanism for propagating overflow errors to the test explicitly, but I'm not sure if its worth the added complexity.

File: lib/inventory/controller.go [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+		Jitter:        utils.NewSeventhJitter(),
+	})
+	defer keepAliveInterval.Stop()
+
+	for {
+		select {
+		case msg := <-handle.Recv():
+			switch m := msg.(type) {
+			case proto.UpstreamInventoryHello:
+				log.Warnf("Unexpected upstream hello on control stream of server %q.", handle.Hello().ServerID)
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/inventory/controller.go:164]: I considered it, but didn't bother adding it since I couldn't think of any useful controller-level fields worth adding. Seemed to me that the ServerID associated with the specific stream was the "key" piece of information, and I didn't want to pre-allocate potentially many thousands of logger instances for that.  Decided it was simpler to just mention the ServerID of the associated stream in all the log lines.

File: lib/inventory/controller.go [hunk 22]
```diff
@@ -0,0 +1,302 @@
...
+	}
+	delete(handle.pings, msg.ID)
+}
+
+func (c *Controller) handlePingRequest(handle *upstreamHandle, req pingRequest) error {
+	handle.pingCounter++
+	ping := proto.DownstreamInventoryPing{
+		ID: handle.pingCounter,
+	}
+	start := time.Now()
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/inventory/controller.go:216]: This is about measuring relative time, so I don't think there'd be any benefit.

File: lib/inventory/controller.go [hunk 23]
```diff
@@ -0,0 +1,302 @@
...
+	}
+	handle.pings[handle.pingCounter] = pendingPing{
+		start: start,
+		rspC:  req.rspC,
+	}
+	return nil
+}
+
+func (c *Controller) handleHeartbeat(handle *upstreamHandle, hb proto.InventoryHeartbeat) error {
+	if hb.SSHServer != nil {
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/inventory/controller.go:231]: My thinking here is that we'll be adding new fields to the heartbeat message over time, and we'll want to add a separate `if hb.FieldName != nil { ... }` block for each one.

File: lib/srv/heartbeatv2_test.go (hunk 11 above)
Comment (fspmarshall) [reply lib/srv/heartbeatv2_test.go:171]: See https://github.com/gravitational/teleport/pull/12715#discussion_r893090454 for my thoughts on this.  Tl;dr is that I'm leaning toward the position that explicit `Close` is a better pattern unless there is a specific reason why the component in question is more suited to implicit closure.  Open to the idea that I'm way off base on that though.  Tell me what you think.

**Review by rosstimothy — COMMENTED** [review]
Please don't squash review fixes (at least before merging). It makes it hard to figure out what you changed since the initial review, especially after a few days.

File: api/client/inventory.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (fspmarshall) [reply api/client/inventory.go:393]: Sorry, what I was trying and failing to express was that the method doesn't actually return an error... we just have an error return type to satisfy an interface.

Plus, even if it did return an error, I'm not sure what we'd do with it here.  We almost always ignore the errors returned from `Close()` calls, and I don't think we care if something goes wrong with the stream if we're closing it anyway.

File: lib/inventory/controller.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/inventory/controller.go:231]: Good point.  Changed `handleHeartbeat` to delegate the server resource to a `handleSSHServerHB` method.

**Review by espadolini — APPROVED** [review]

File: api/client/inventory.go [hunk 24]
```diff
@@ -0,0 +1,511 @@
...
+	"sync"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/trace"
+	"github.com/gravitational/trace/trail"
+)
+
+// DownstreamInventoryControlStream is the client/agent side of a bidirectional stream established
+// between teleport instances and auth servers.
+type DownstreamInventoryControlStream interface {
```
Comment (zmb3) [inline api/client/inventory.go:31]: I won't block the PR on this, but this inventory.go file seems like it should be pretty easy to write some unit tests to verify the basic functionality (start recv loops, send a ping, close and wait for termination, etc)

File: api/client/inventory.go [hunk 25]
```diff
@@ -0,0 +1,511 @@
...
+
+func (i *upstreamICS) Close() error {
+	return i.CloseWithError(nil)
+}
+
+func (i *upstreamICS) CloseWithError(err error) error {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	if i.err != nil {
+		// already closed
```
Comment (zmb3) [inline api/client/inventory.go:495]: Can you document this behavior in the godoc?

```
// CloseWithError closes the stream and records the error
// which can later be retrieved with Error(). The error is
// only recorded on the first call - subseqeuent calls to
// CloseWithError are no ops.
```

File: lib/auth/auth_with_roles.go [hunk 26]
```diff
@@ -610,18 +610,167 @@ func (a *ServerWithRoles) GenerateHostCerts(ctx context.Context, req *proto.Host
...
 	}
+
 	existingRoles, err := types.NewTeleportRoles(a.context.User.GetRoles())
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	// prohibit privilege escalations through role changes
-	if !existingRoles.Equals(types.SystemRoles{req.Role}) {
+	if req.Role == types.RoleInstance {
```
Comment (zmb3) [inline lib/auth/auth_with_roles.go:619]: There are a lot of different code paths through here with all the looping and branching - I would be more comfortable with some tests to exercise this functionality.

File: lib/auth/auth.go [hunk 27]
```diff
@@ -2225,6 +2250,49 @@ func (a *Server) GenerateHostCerts(ctx context.Context, req *proto.HostCertsRequ
 	}, nil
 }
 
+func (a *Server) UnstableAssertSystemRole(ctx context.Context, req proto.UnstableSystemRoleAssertion) error {
```
Comment (zmb3) [inline lib/auth/auth.go:2253]: Can you add godocs to this? `UnstableAssertSystemRole` sounds scary to someone not familiar with what's going on.

File: lib/auth/auth.go [hunk 28]
```diff
@@ -2225,6 +2250,49 @@ func (a *Server) GenerateHostCerts(ctx context.Context, req *proto.HostCertsRequ
...
+	}
+
+	return proto.InventoryPingResponse{
+		Duration: d,
+	}, nil
+}
+
+// TokenExpiredOrNotFound is a special message returned by the auth server when provisioning
+// tokens are either past their TTL, or could not be found.
+const TokenExpiredOrNotFound = "token expired or not found"
```
[the remark this answers is not in the document]
Comment (espadolini) [reply lib/auth/auth.go:2294]: That should even work through the grpc error conversion provided by `trail`, right?

File: lib/utils/retry.go [hunk 29]
```diff
@@ -33,6 +33,11 @@ import (
 // repeated calls.
 var HalfJitter = NewHalfJitter()
 
+// SeventhJitter is a global jitter instance used for one-off jitters.
+// Prefer instantiating a new jitter instance for operations that require
+// repeated calls.
+var SeventhJitter = NewSeventhJitter()
```
Comment (zmb3) [inline lib/utils/retry.go:39]: Is this used anywhere?

File: lib/utils/retry.go [hunk 30]
```diff
@@ -155,6 +160,27 @@ func NewConstant(interval time.Duration) (*Linear, error) {
...
 }
 
+// NewDefaultLinear creates a linear retry using a half jitter, 10s step, and maxing out
+// at 1 minute. These values were selected by reviewing commonly used parameters elsewhere
+// in the code base, which (at the time of writing) seem to converge on approximately this
+// configuration for "critical but potentially load-inducing" operations like cache watcher
+// registration and auth connector setup. It also includes an auto-reset value of 5m. Auto-reset
+// is less commonly used, and if used should probably be shorter, but 5m is a reasonable
+// safety net to reduce the impact of accidental misuse.
+func NewDefaultLinear() *Linear {
```
Comment (zmb3) [inline lib/utils/retry.go:170]: Maybe add a TODO to actually use this throughout the codebase? Otherwise it's not a default, it's just the one used by the inventory control streams.

File: tool/tctl/common/inventory_command.go [hunk 31]
```diff
@@ -0,0 +1,108 @@
...
+	getConnected bool
+
+	inventoryStatus *kingpin.CmdClause
+	inventoryPing   *kingpin.CmdClause
+}
+
+// Initialize allows AccessRequestCommand to plug itself into the CLI parser
+func (c *InventoryCommand) Initialize(app *kingpin.Application, config *service.Config) {
+	c.config = config
+	inventory := app.Command("inventory", "Manage teleport instance inventory").Hidden()
```
Comment (zmb3) [inline tool/tctl/common/inventory_command.go:48]:
```suggestion
	inventory := app.Command("inventory", "Manage Teleport instance inventory").Hidden()
```

File: tool/tctl/main.go [hunk 32]
```diff
@@ -42,6 +42,7 @@ func main() {
 		&common.AccessCommand{},
 		&common.LockCommand{},
 		&common.BotsCommand{},
+		&common.InventoryCommand{},
```
Comment (zmb3) [inline tool/tctl/main.go:45]: Don't forget to put this in package main for `e` as well.

File: lib/utils/interval/interval.go [hunk 33]
```diff
@@ -129,6 +140,15 @@ func (i *Interval) run(timer *time.Timer) {
 			// output channel is set.
 			timer.Reset(i.duration())
 			ch = i.ch
+		case <-i.reset:
+			// stop and drain timer
+			if !timer.Stop() {
+				<-timer.C
+			}
```
Comment (zmb3) [inline lib/utils/interval/interval.go:147]: My brain is kind of fried after reading through all of this. I can't say for sure whether this is correct or not, but looks like a particularly tricky spot. Are you confident we got this right here? ( timer stopped twice, drain necessary, etc)

**Review by zmb3 — APPROVED** [review]

**Comment by r0mant** [comment]
@fspmarshall @zmb3 @espadolini @rosstimothy Is this PR good to merge? Would like to get this in before we start testing v10. If extra integration tests are needed, those can go in a follow up PR.

File: lib/auth/auth_with_roles.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/auth/auth_with_roles.go]: Checks are mutually exclusive, so it still needs to be an if/else, but I factored the body out to a separate helper function, so much more readable now.

File: lib/auth/auth.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/auth/auth.go:2305]: I feel like I tried this in the past, and `Is` didn't work across the network.  Maybe it does now, but IMO its still unnecessarily brittle.

File: lib/inventory/controller.go [hunk 34]
```diff
@@ -0,0 +1,315 @@
...
+	auth            Auth
+	serverKeepAlive time.Duration
+	serverTTL       time.Duration
+	testEvents      chan testEvent
+	closeContext    context.Context
+	cancel          context.CancelFunc
+}
+
+// NewController sets up a new controller instance.
+func NewController(auth Auth, opts ...controllerTestOption) *Controller {
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/inventory/controller.go]: Haha, alright.  I've changed them to public.  I still prefer options to config here since there's only one public parameter at the moment, but the option type is now public and the options struct has a `SetDefaults` method so that adding new public inputs in the future will be easy and won't change the type signature.

File: lib/inventory/controller.go [hunk 35]
```diff
@@ -0,0 +1,315 @@
...
+	if handle.sshServerLease != nil {
+		lease := *handle.sshServerLease
+		lease.Expires = time.Now().Add(c.serverTTL).UTC()
+		if err := c.auth.KeepAliveServer(c.closeContext, lease); err != nil {
+			c.testEvent(sshKeepAliveErr)
+			handle.sshServerKeepAliveErrs++
+
+			// fail on third error. this is arbitrary, but feels reasonable since if we're on our
+			// third error, 3-4m have gone by, so the problem is almost certainly persistent.
+			shouldClose := handle.sshServerKeepAliveErrs > 2
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/inventory/controller.go]: Added it to `options.SetDefault`.  Didn't add a public setter since we don't currently override this anywhere, but it'll now be super-simple to start overriding it in the future if we want.

File: api/client/inventory.go (hunk 24 above)
Comment (fspmarshall) [reply api/client/inventory.go:31]: Added test in `lib/auth` that sets up a client and performs basic hello exchange + ping.

File: lib/auth/auth_with_roles.go (hunk 26 above)
Comment (fspmarshall) [reply lib/auth/auth_with_roles.go]: Added unit test for happy path, and for attempting to request system roles without authorization.

File: lib/utils/retry.go (hunk 29 above)
Comment (fspmarshall) [reply lib/utils/retry.go:39]: Currently just in `srv/heartbeatv2.go` currently, but seemed worth creating the global for it since we get a lot of use out of the corresponding `utils.HalfJitter`.

File: lib/utils/interval/interval.go (hunk 33 above)
Comment (fspmarshall) [reply lib/utils/interval/interval.go:147]: This method of stop/drain is lifted verbatim from the official docs for how to handle stopping a timer in a concurrent context where you aren't certain if it has fired and/or been drained.
